iceoryx_utils/include/iceoryx_utils/cxx/unique_ptr.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_UTILS_CXX_UNIQUE_PTR_HPP
+#define IOX_UTILS_CXX_UNIQUE_PTR_HPP
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+
+#include <functional>
+
+namespace iox
+{
+namespace cxx
+{
+///
+/// Only a subset of STL's 'std::unique_ptr' has been implemented so far.
+/// Main difference is that 'Deleter' is a function reference ('cxx:function_ref') here compared to a 'std::function' in
+/// the STL implementation.
+///
+template <typename T>
+class unique_ptr
+{
+  public:
+    using ptr_t = T*;
+
+    unique_ptr() = delete;
+
+    ///
+    /// @brief unique_ptr Creates an empty unique ptr that owns nothing. Can be passed ownership later via reset.
+    ///
+    unique_ptr(function_ref<void(ptr_t)>&& deleter) noexcept;
+
+    ///
+    /// @brief unique_ptr Creates a unique pointer that takes ownership of an object.
+    /// @details A deleter must always be provided as no default can be provided given that no heap is used.
+    /// The unique_ptr must know how to delete the managed object when the pointer goes out of scope.
+    /// @param ptr The raw pointer to the object to be managed.
+    /// @param deleter The deleter function for cleaning up the managed object. As cxx:function_ref used for the deleter
+    ///                is non-owning the user needs to care about the lifetime of the callable!
+    ///
+    unique_ptr(ptr_t ptr, function_ref<void(ptr_t)>&& deleter) noexcept;
```

Comment:
Const correctness. Please write `unique_ptr(ptr_t const ptr...` since `ptr_t` is in the end `T*`